File: sdks/python/apache_beam/runners/direct_runner.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""DirectPipelineRunner, executing on the local machine.

The DirectPipelineRunner class implements what is called in Dataflow
parlance the "direct runner". Such a runner executes the entire graph
of transformations belonging to a pipeline on the local machine.
"""

from __future__ import absolute_import

import collections
import itertools
import logging

from apache_beam import coders
from apache_beam import error
from apache_beam.io import fileio
from apache_beam.io import iobase
from apache_beam.pvalue import DictPCollectionView
from apache_beam.pvalue import EmptySideInput
from apache_beam.pvalue import IterablePCollectionView
from apache_beam.pvalue import ListPCollectionView
from apache_beam.pvalue import SingletonPCollectionView
from apache_beam.runners.common import DoFnRunner
from apache_beam.runners.common import DoFnState
from apache_beam.runners.runner import PipelineResult
from apache_beam.runners.runner import PipelineRunner
from apache_beam.runners.runner import PipelineState
from apache_beam.runners.runner import PValueCache
from apache_beam.transforms.window import GlobalWindows
from apache_beam.transforms.window import WindowedValue
from apache_beam.typehints.typecheck import OutputCheckWrapperDoFn
from apache_beam.typehints.typecheck import TypeCheckError
from apache_beam.typehints.typecheck import TypeCheckWrapperDoFn
from apache_beam.utils import counters
from apache_beam.utils.options import TypeOptions


class DirectPipelineRunner(PipelineRunner):
  """A local pipeline runner.

  The runner computes everything locally and does not make any attempt to
  optimize for time or space.
  """

  def __init__(self, cache=None):
    # Cache of values computed while the runner executes a pipeline.
    self._cache = cache if cache is not None else PValueCache()
    self._counter_factory = counters.CounterFactory()
    # Element counts used for debugging footprint issues in the direct runner.
    # The values computed are used only for logging and do not take part in
    # any decision making logic. The key for the counter dictionary is either
    # the full label for the transform producing the elements or a tuple
    # (full label, output tag) for ParDo transforms since they can output values
    # on multiple outputs.
    self.debug_counters = {}
    self.debug_counters['element_counts'] = collections.Counter()

  @property
  def cache(self):
    return self._cache

  def get_pvalue(self, pvalue):
    """Gets the PValue's computed value from the runner's cache."""
    try:
      return self._cache.get_pvalue(pvalue)
    except KeyError:
      raise error.PValueError('PValue is not computed.')

  def clear_pvalue(self, pvalue):
    """Removes a PValue from the runner's cache."""
    self._cache.clear_pvalue(pvalue)

  def skip_if_cached(func):  # pylint: disable=no-self-argument
    """Decorator to skip execution of a transform if value is cached."""

    def func_wrapper(self, pvalue, *args, **kwargs):
      logging.debug('Current: Debug counters: %s', self.debug_counters)
      if self._cache.is_cached(pvalue):  # pylint: disable=protected-access
        return
      else:
        func(self, pvalue, *args, **kwargs)
    return func_wrapper

  def run(self, pipeline):
    super(DirectPipelineRunner, self).run(pipeline)
    logging.info('Final: Debug counters: %s', self.debug_counters)
    return DirectPipelineResult(state=PipelineState.DONE,
                                counter_factory=self._counter_factory)

  @skip_if_cached
  def run_CreatePCollectionView(self, transform_node):
    transform = transform_node.transform
    view = transform.view
    values = self._cache.get_pvalue(transform_node.inputs[0])
    if isinstance(view, SingletonPCollectionView):
      has_default, default_value = view._view_options()  # pylint: disable=protected-access
      if len(values) == 0:
        if has_default:
          result = default_value
        else:
          result = EmptySideInput()
      elif len(values) == 1:
        # TODO(ccy): Figure out whether side inputs should ever be given as
        # windowed values
        result = values[0].value
      else:
        raise ValueError(('PCollection with more than one element accessed as '
                          'a singleton view: %s.') % view)
    elif isinstance(view, IterablePCollectionView):
      result = [v.value for v in values]
    elif isinstance(view, ListPCollectionView):
      result = [v.value for v in values]
    elif isinstance(view, DictPCollectionView):
      result = dict(v.value for v in values)
    else:
      raise NotImplementedError

    self._cache.cache_output(transform_node, result)

  @skip_if_cached
  def run_ParDo(self, transform_node):
    transform = transform_node.transform

    side_inputs = [self._cache.get_pvalue(view)
                   for view in transform_node.side_inputs]

    # TODO(robertwb): Do this type checking inside DoFnRunner to get it on
    # remote workers as well?
    options = transform_node.inputs[0].pipeline.options
    if options is not None and options.view_as(TypeOptions).runtime_type_check:
      transform.dofn = TypeCheckWrapperDoFn(
          transform.dofn, transform.get_type_hints())

    # TODO(robertwb): Should this be conditionally done on the workers as well?
    transform.dofn = OutputCheckWrapperDoFn(
        transform.dofn, transform_node.full_label)

    class RecordingReceiverSet(object):

      def __init__(self, tag):
        self.tag = tag

      def output(self, element):
        results[self.tag].append(element)

    class TaggedReceivers(dict):

      def __missing__(self, key):
        return RecordingReceiverSet(key)

    results = collections.defaultdict(list)
    # Some tags may be empty.
    for tag in transform.side_output_tags:
      results[tag] = []

    runner = DoFnRunner(transform.dofn, transform.args, transform.kwargs,
                        side_inputs, transform_node.inputs[0].windowing,
                        tagged_receivers=TaggedReceivers(),
                        step_name=transform_node.full_label,
                        state=DoFnState(self._counter_factory))
    runner.start()
    for v in self._cache.get_pvalue(transform_node.inputs[0]):
      runner.process(v)
    runner.finish()

    self._cache.cache_output(transform_node, [])
    for tag, value in results.items():
      self.debug_counters['element_counts'][
          (transform_node.full_label, tag)] += len(value)
      self._cache.cache_output(transform_node, tag, value)

  @skip_if_cached
  def run_GroupByKeyOnly(self, transform_node):
    result_dict = collections.defaultdict(list)
    # The input type of a GroupByKey will be KV[Any, Any] or more specific.
    kv_type_hint = transform_node.transform.get_type_hints().input_types[0]
    key_coder = coders.registry.get_coder(kv_type_hint[0].tuple_types[0])

    for wv in self._cache.get_pvalue(transform_node.inputs[0]):
      if (isinstance(wv, WindowedValue) and
          isinstance(wv.value, collections.Iterable) and len(wv.value) == 2):
        k, v = wv.value
        # We use as key a string encoding of the key object to support keys
        # that are based on custom classes. This mimics also the remote
        # execution behavior where key objects are encoded before being written
        # to the shuffler system responsible for grouping.
        result_dict[key_coder.encode(k)].append(v)
      else:
        raise TypeCheckError('Input to GroupByKeyOnly must be a PCollection of '
                             'windowed key-value pairs. Instead received: %r.'
                             % wv)

    gbk_result = map(
        GlobalWindows.windowed_value,
        ((key_coder.decode(k), v) for k, v in result_dict.iteritems()))
    self.debug_counters['element_counts'][
        transform_node.full_label] += len(gbk_result)
    self._cache.cache_output(transform_node, gbk_result)

  @skip_if_cached
  def run_Create(self, transform_node):
    transform = transform_node.transform
    create_result = [GlobalWindows.windowed_value(v) for v in transform.value]
    self.debug_counters['element_counts'][
        transform_node.full_label] += len(create_result)
    self._cache.cache_output(transform_node, create_result)

  @skip_if_cached
  def run_Flatten(self, transform_node):
    flatten_result = list(
        itertools.chain.from_iterable(
            self._cache.get_pvalue(pc) for pc in transform_node.inputs))
    self.debug_counters['element_counts'][
        transform_node.full_label] += len(flatten_result)
    self._cache.cache_output(transform_node, flatten_result)

  @skip_if_cached
  def run_Read(self, transform_node):
    # TODO(chamikara) Implement a more generic way for passing PipelineOptions
    # to sources and sinks when using DirectRunner.
    source = transform_node.transform.source
    source.pipeline_options = transform_node.inputs[0].pipeline.options

    def read_values(reader):
      read_result = [GlobalWindows.windowed_value(e) for e in reader]
      self.debug_counters['element_counts'][
          transform_node.full_label] += len(read_result)
      self._cache.cache_output(transform_node, read_result)

    if isinstance(source, iobase.BoundedSource):
      # Getting a RangeTracker for the default range of the source and reading
      # the full source using that.
      range_tracker = source.get_range_tracker(None, None)
      reader = source.read(range_tracker)
      read_values(reader)
    else:
      with source.reader() as reader:
        read_values(reader)

  @skip_if_cached
  def run__NativeWrite(self, transform_node):
    sink = transform_node.transform.sink
    if isinstance(sink, fileio.NativeTextFileSink):
      assert sink.num_shards in (0, 1)
      if sink.shard_name_template:
        sink.file_path += '-00000-of-00001'
      sink.file_path += sink.file_name_suffix
    sink.pipeline_options = transform_node.inputs[0].pipeline.options
    with sink.writer() as writer:
      for v in self._cache.get_pvalue(transform_node.inputs[0]):
        self.debug_counters['element_counts'][transform_node.full_label] += 1
        writer.Write(v.value)


class DirectPipelineResult(PipelineResult):
  """A DirectPipelineResult provides access to info about a pipeline."""

  def __init__(self, state, counter_factory=None):
    super(DirectPipelineResult, self).__init__(state)
    self._counter_factory = counter_factory

  def aggregated_values(self, aggregator_or_name):
    return self._counter_factory.get_aggregator_values(aggregator_or_name)


class EagerPipelineRunner(DirectPipelineRunner):

  is_eager = True

  def __init__(self):
    super(EagerPipelineRunner, self).__init__()
    self._seen_transforms = set()

  def run_transform(self, transform):
    if transform not in self._seen_transforms:
      self._seen_transforms.add(transform)
      super(EagerPipelineRunner, self).run_transform(transform)


class DiskCachedPipelineRunner(DirectPipelineRunner):
  """A DirectPipelineRunner that uses a disk backed cache.

  DiskCachedPipelineRunner uses a temporary disk backed cache for running
  pipelines. This allows for running pipelines that will require more memory
  than it is available, however this comes with a performance cost due to disk
  IO.

  Memory requirement for DiskCachedPipelineRunner is approximately capped by the
  single transform in the pipeline that consumes and outputs largest total
  collection (i.e. inputs, side-inputs and outputs in aggregate). In the extreme
  case a where a transform will use all previous intermediate values as input,
  memory requirements for DiskCachedPipelineRunner will be the same as
  DirectPipelineRunner.
  """

  def __init__(self):
    self._null_cache = ()
    super(DiskCachedPipelineRunner, self).__init__(self._null_cache)

  def run(self, pipeline):
    try:
      self._cache = PValueCache(use_disk_backed_cache=True)
      return super(DirectPipelineRunner, self).run(pipeline)
    finally:
      del self._cache
      self._cache = self._null_cache

  @property
  def cache(self):
    raise NotImplementedError(
        'DiskCachedPipelineRunner does not keep cache outside the scope of its '
        'run method.')

File: sdks/typescript/src/apache_beam/runners/direct_runner.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as uuid from "uuid";
import * as protobufjs from "protobufjs";
import Long from "long";

import * as runnerApi from "../proto/beam_runner_api";
import * as fnApi from "../proto/beam_fn_api";
import { PTransform, PCollection } from "../proto/beam_runner_api";
import { ProcessBundleDescriptor } from "../proto/beam_fn_api";
import { JobState_Enum } from "../proto/beam_job_api";

import { Pipeline } from "../internal/pipeline";
import { Root } from "../pvalue";
import { impulse, groupByKey } from "../transforms/internal";
import { Runner, PipelineResult } from "./runner";
import * as worker from "../worker/worker";
import * as operators from "../worker/operators";
import { createStateKey } from "../worker/pardo_context";
import * as metrics from "../worker/metrics";
import * as state from "../worker/state";
import { parDo } from "../transforms/pardo";
import {
  Window,
  GlobalWindow,
  Instant,
  PaneInfo,
  WindowedValue,
} from "../values";
import { PaneInfoCoder } from "../coders/standard_coders";
import { Coder, Context as CoderContext } from "../coders/coders";
import * as environments from "../internal/environments";
import { serializeFn, deserializeFn } from "../internal/serialize";

const SUPPORTED_REQUIREMENTS: string[] = [];

export function directRunner(options: Object = {}): Runner {
  return new DirectRunner(options);
}

class DirectRunner extends Runner {
  // All the operators for a given pipeline should share the same state.
  // This global mapping allows operators to look up a shared state object for
  // a given pipeline on deserialization.
  static inMemoryStatesRefs: Map<string, InMemoryStateProvider> = new Map();

  constructor(private options: Object = {}) {
    super();
  }

  unsupportedFeatures(pipeline, options: Object = {}): string[] {
    return [...this.unsupportedFeaturesIter(pipeline, options)];
  }

  *unsupportedFeaturesIter(proto: runnerApi.Pipeline, options: Object = {}) {
    for (const requirement of proto.requirements) {
      if (!SUPPORTED_REQUIREMENTS.includes(requirement)) {
        yield requirement;
      }
    }

    for (const env of Object.values(proto.components!.environments)) {
      if (
        env.urn &&
        env.urn !== environments.TYPESCRIPT_DEFAULT_ENVIRONMENT_URN
      ) {
        yield env.urn;
      }
    }

    for (const windowing of Object.values(
      proto.components!.windowingStrategies,
    )) {
      if (
        ![
          runnerApi.MergeStatus_Enum.UNSPECIFIED,
          runnerApi.MergeStatus_Enum.NON_MERGING,
          runnerApi.MergeStatus_Enum.ALREADY_MERGED,
        ].includes(windowing.mergeStatus)
      ) {
        yield "MergeStatus=" + windowing.mergeStatus;
      }
      if (windowing.outputTime !== runnerApi.OutputTime_Enum.END_OF_WINDOW) {
        yield "OutputTime=" + windowing.outputTime;
      }
    }
  }

  async runPipeline(p: runnerApi.Pipeline): Promise<PipelineResult> {
    const stateProvider = new InMemoryStateProvider();
    const stateCacheRef = uuid.v4();
    DirectRunner.inMemoryStatesRefs.set(stateCacheRef, stateProvider);

    try {
      const proto = rewriteSideInputs(p, stateCacheRef);
      const descriptor: ProcessBundleDescriptor = {
        id: "",
        transforms: proto.components!.transforms,
        pcollections: proto.components!.pcollections,
        windowingStrategies: proto.components!.windowingStrategies,
        coders: proto.components!.coders,
        environments: proto.components!.environments,
      };

      const processor = new worker.BundleProcessor(
        descriptor,
        null!,
        new state.CachingStateProvider(stateProvider),
        [impulse.urn],
      );
      await processor.process("bundle_id");

      return new (class DirectPipelineResult extends PipelineResult {
        waitUntilFinish(duration?: number) {
          return Promise.resolve(JobState_Enum.DONE);
        }
        async rawMetrics() {
          const shortIdCache = new metrics.MetricsShortIdCache();
          const monitoringData = processor.monitoringData(shortIdCache);
          return Array.from(monitoringData.entries()).map(([id, payload]) =>
            shortIdCache.asMonitoringInfo(id, payload),
          );
        }
      })();
    } finally {
      DirectRunner.inMemoryStatesRefs.delete(stateCacheRef);
    }
  }
}

// Only to be used in direct runner, as this will fire an element per worker, not per pipeline.
class DirectImpulseOperator implements operators.IOperator {
  receiver: operators.Receiver;

  constructor(
    public transformId: string,
    transform: PTransform,
    context: operators.OperatorContext,
  ) {
    this.receiver = context.getReceiver(
      onlyElement(Object.values(transform.outputs)),
    );
  }

  process(wvalue: WindowedValue<unknown>) {
    return operators.NonPromise;
  }

  async startBundle() {
    this.receiver.receive({
      value: new Uint8Array(),
      windows: [new GlobalWindow()],
      pane: PaneInfoCoder.ONE_AND_ONLY_FIRING,
      timestamp: Long.fromValue("-9223372036854775"), // TODO: (Cleanup) Pull constant out of proto, or at least as a constant elsewhere.
    });
  }

  async finishBundle() {}
}

operators.registerOperator(impulse.urn, DirectImpulseOperator);

// Only to be used in direct runner, as this will only group within a single bundle.
// TODO: (Extension) This could be used as a base for the PGBKOperation operator,
// and then this class could simply invoke that with an unbounded size and the
// concat-to-list CombineFn.
class DirectGbkOperator implements operators.IOperator {
  receiver: operators.Receiver;
  groups: Map<string, unknown[]>;
  keyCoder: Coder<unknown>;
  windowCoder: Coder<Window>;

  constructor(
    public transformId: string,
    transform: PTransform,
    context: operators.OperatorContext,
  ) {
    this.receiver = context.getReceiver(
      onlyElement(Object.values(transform.outputs)),
    );
    const inputPc =
      context.descriptor.pcollections[
        onlyElement(Object.values(transform.inputs))
      ];
    this.keyCoder = context.pipelineContext.getCoder(
      context.descriptor.coders[inputPc.coderId].componentCoderIds[0],
    );
    const windowingStrategy =
      context.descriptor.windowingStrategies[inputPc.windowingStrategyId];
    if (
      windowingStrategy.mergeStatus !== runnerApi.MergeStatus_Enum.NON_MERGING
    ) {
      throw new Error("Unsupported non-merging WindowFn: " + windowingStrategy);
    }
    if (
      windowingStrategy.outputTime !== runnerApi.OutputTime_Enum.END_OF_WINDOW
    ) {
      throw new Error(
        "Unsupported windowing output time: " + windowingStrategy,
      );
    }
    this.windowCoder = context.pipelineContext.getCoder(
      windowingStrategy.windowCoderId,
    );
  }

  process(wvalue: WindowedValue<any>) {
    for (const window of wvalue.windows) {
      const wkey =
        operators.encodeToBase64(window, this.windowCoder) +
        " " +
        operators.encodeToBase64(wvalue.value.key, this.keyCoder);
      if (!this.groups.has(wkey)) {
        this.groups.set(wkey, []);
      }
      this.groups.get(wkey)!.push(wvalue.value.value);
    }
    return operators.NonPromise;
  }

  async startBundle() {
    this.groups = new Map();
  }

  async finishBundle() {
    for (const [wkey, values] of this.groups) {
      const parts = wkey.split(" ");
      const encodedWindow = parts[0];
      const encodedKey = parts[1];
      const window = operators.decodeFromBase64(
        encodedWindow,
        this.windowCoder,
      );
      const maybePromise = this.receiver.receive({
        value: {
          key: operators.decodeFromBase64(encodedKey, this.keyCoder),
          value: values,
        },
        windows: [window],
        timestamp: window.maxTimestamp(),
        pane: PaneInfoCoder.ONE_AND_ONLY_FIRING,
      });
      if (maybePromise !== operators.NonPromise) {
        await maybePromise;
      }
    }
    this.groups = null!;
  }
}

operators.registerOperator(groupByKey.urn, DirectGbkOperator);

/**
 * Rewrites the pipeline to be suitable for running as a single "bundle."
 *
 * Given a pipeline of the form
 *
 *    ...          ...
 *     |            |
 * mainPColl    sidePColl
 *     |            |
 *   parDo <-------/
 *     |
 *    ...
 *
 * Rewrites this to be in the form.
 *
 *    ...          ...
 *     |            |
 * mainPColl    sidePColl
 *     |            |
 *     |      collectSideOp
 *     |            |
 *     |        emptyPColl
 *     | <- - - - -/
 *     |
 *  bufferOp
 *     |
 * mainPcollCopy
 *     |
 *   parDo <----- sidePCollCopy
 *
 * Where collectSideOp collects its inputs for future retrieval and bufferOp
 * buffers all its inputs, invoking the downstream parDo only once finishBundle
 * is called. The dotted line from collectSideOp to bufferOp is not used to
 * pass any data, but acts as a control edge forcing bufferOp to be finished
 * only after all side inputs have been fully collected.
 *
 * A sidePCollCopy is also placed in the graph to allow for lookup of metadata
 * when executing the parDo operaton.
 */
function rewriteSideInputs(p: runnerApi.Pipeline, pipelineStateRef: string) {
  function uniqueName(container, base) {
    let name = base;
    let counter = 0;
    while (container[name]) {
      name = base + counter++;
    }
    return name;
  }

  p = runnerApi.Pipeline.clone(p);
  const pcolls = p.components!.pcollections;
  const transforms = p.components!.transforms;
  for (const [transformId, transform] of Object.entries(transforms)) {
    if (
      transform.spec?.urn === parDo.urn &&
      Object.keys(transform.inputs).length > 1
    ) {
      const spec = runnerApi.ParDoPayload.fromBinary(transform.spec!.payload);

      // Figure out which input is the main input.
      // TODO: (Typescript) Is there a clean way to do a set difference?
      // There is a proposal for Set.prototype.intersect: https://github.com/tc39/proposal-set-methods
      // But I don't think there is a nicer way at the moment.
      let mainPCollTag: string = undefined!;
      for (const tag of Object.keys(transform.inputs)) {
        if (!spec.sideInputs[tag]) {
          if (mainPCollTag) throw new Error("Multiple non-side inputs.");
          mainPCollTag = tag;
        }
      }
      if (!mainPCollTag) throw new Error("Missing non-side input.");

      // Add the extra logic for each of the side inputs.
      const bufferInputs = { mainPCollTag: transform.inputs[mainPCollTag] };
      for (const side of Object.keys(spec.sideInputs)) {
        const sidePCollId = transform.inputs[side];
        const sideCopyId = uniqueName(pcolls, sidePCollId + "-" + side);
        pcolls[sideCopyId] = pcolls[sidePCollId];
        transform.inputs[side] = sideCopyId;
        const controlPCollId = uniqueName(
          pcolls,
          sidePCollId + "-" + side + "-control",
        );
        pcolls[controlPCollId] = pcolls[transform.inputs[mainPCollTag]];
        bufferInputs[side] = controlPCollId;
        const collectTransformId = uniqueName(
          transforms,
          transformId + "-" + side + "-collect",
        );
        transforms[collectTransformId] = runnerApi.PTransform.create({
          spec: {
            urn: CollectSideOperator.urn,
            payload: serializeFn({
              transformId: transformId,
              sideInputId: side,
              accessPattern: spec.sideInputs[side].accessPattern!.urn,
              pipelineStateRef: pipelineStateRef,
            }),
          },
          inputs: { input: sidePCollId },
          outputs: { out: controlPCollId },
        });
      }

      // Add the buffering operator.
      const bufferId = uniqueName(transforms, transformId + "-buffer");
      const bufferOutputId = uniqueName(pcolls, bufferId + "-out");
      pcolls[bufferOutputId] = pcolls[mainPCollTag];
      transforms[bufferId] = runnerApi.PTransform.create({
        spec: { urn: BufferOperator.urn, payload: new Uint8Array() },
        inputs: bufferInputs,
        outputs: { out: bufferOutputId },
      });
      transform.inputs[mainPCollTag] = bufferOutputId;
    }
  }

  return p;
}

class CollectSideOperator implements operators.IOperator {
  static urn = "beam:transforms:node_ts_direct:collect_side:v1";

  stateProvider: InMemoryStateProvider;

  parDoTransformId: string;
  accessPattern: string;
  sideInputId: string;

  receiver: operators.Receiver;
  windowCoder: Coder<Window>;
  elementCoder: Coder<unknown>;

  constructor(
    public transformId: string,
    transform: PTransform,
    context: operators.OperatorContext,
  ) {
    this.receiver = context.getReceiver(
      onlyElement(Object.values(transform.outputs)),
    );
    const payload = deserializeFn(transform.spec!.payload!);
    this.parDoTransformId = payload.transformId;
    this.accessPattern = payload.accessPattern;
    this.sideInputId = payload.sideInputId;
    this.stateProvider = DirectRunner.inMemoryStatesRefs.get(
      payload.pipelineStateRef,
    )!;

    const inputPc =
      context.descriptor.pcollections[
        onlyElement(Object.values(transform.inputs))
      ];
    this.elementCoder = context.pipelineContext.getCoder(inputPc.coderId);
    const windowingStrategy =
      context.descriptor.windowingStrategies[inputPc.windowingStrategyId];
    this.windowCoder = context.pipelineContext.getCoder(
      windowingStrategy.windowCoderId,
    );
  }

  async startBundle() {}

  process(wvalue: WindowedValue<unknown>) {
    for (const window of wvalue.windows) {
      const writer = new protobufjs.Writer();
      this.elementCoder.encode(
        wvalue.value,
        writer,
        CoderContext.needsDelimiters,
      );
      const encodedElement = writer.finish();
      this.stateProvider.appendState(
        createStateKey(
          this.parDoTransformId,
          this.accessPattern,
          this.sideInputId,
          window,
          this.windowCoder,
        ),
        encodedElement,
      );
    }
    return operators.NonPromise;
  }

  async finishBundle() {}
}

operators.registerOperator(CollectSideOperator.urn, CollectSideOperator);

class BufferOperator implements operators.IOperator {
  static urn = "beam:transforms:node_ts_direct:buffer:v1";
  receiver: operators.Receiver;
  elements: WindowedValue<unknown>[];

  constructor(
    public transformId: string,
    transform: PTransform,
    context: operators.OperatorContext,
  ) {
    this.receiver = context.getReceiver(
      onlyElement(Object.values(transform.outputs)),
    );
  }

  process(wvalue: WindowedValue<unknown>) {
    this.elements.push(wvalue);
    return operators.NonPromise;
  }

  async startBundle() {
    this.elements = [];
  }

  async finishBundle() {
    for (const element of this.elements) {
      const maybePromise = this.receiver.receive(element);
      if (maybePromise !== operators.NonPromise) {
        await maybePromise;
      }
    }
  }
}

operators.registerOperator(BufferOperator.urn, BufferOperator);

class InMemoryStateProvider implements state.StateProvider {
  chunks: Map<string, Uint8Array[]> = new Map();

  getState<T>(
    stateKey: fnApi.StateKey,
    decode: (data: Uint8Array) => T,
  ): state.MaybePromise<T> {
    return {
      type: "value",
      value: decode(state.Uint8ArrayConcat(this.getStateEntry(stateKey))),
    };
  }

  appendState(stateKey: fnApi.StateKey, encodedData: Uint8Array) {
    this.getStateEntry(stateKey).push(encodedData);
  }

  getStateEntry<T>(stateKey: fnApi.StateKey) {
    const cacheKey = Buffer.from(fnApi.StateKey.toBinary(stateKey)).toString(
      "base64",
    );
    if (!this.chunks.has(cacheKey)) {
      this.chunks.set(cacheKey, []);
    }
    return this.chunks.get(cacheKey)!;
  }
}

/////

function onlyElement<T>(arg: T[]): T {
  if (arg.length > 1) {
    Error("Expecting exactly one element.");
  }
  return arg[0];
}
